File: Sources/windows/BitTherapyCLR/src/species/species_parser.cpp
#include "species_parser.h"

#include <fstream>
#include <iostream>
#include <optional>
#include <sstream>
#include <vector>

#include "species_model.h"

#include "../utils/utils.h"

#include "../../dependencies/nlohmann_json.hpp"

static const std::string SPECIES_ANIMATION_POSITION_FROM_ENTITY_BOTTOM_LEFT = "fromEntityBottomLeft";

std::optional<Species> SpeciesParser::parseFromFile(const std::string& filePath) const {
    std::ifstream file(filePath);
    if (!file.is_open()) {
        return std::nullopt;
    }
    std::stringstream buffer;
    buffer << file.rdbuf();
    return parse(buffer.str()); 
}

std::optional<Species> SpeciesParser::parse(const std::string& jsonString) const {
    try {
        auto json = nlohmann::json::parse(jsonString);
        Species species(
            json.value("id", "missingno"),
            json.value("speed", 1.0),
            json.value("scale", 1.0)
        );

        species.capabilities = json.value("capabilities", std::vector<std::string>({}));
        species.dragPath = json.value("dragPath", "drag");
        species.movementPath = json.value("movementPath", "walk");
        species.zIndex = json.value("zIndex", 0);

        if (json.contains("animations")) {
            for (const auto& anim : json.at("animations")) {
                species.animations.emplace_back(
                    anim.value("id", "undefined"),
                    anim.value("position", SPECIES_ANIMATION_POSITION_FROM_ENTITY_BOTTOM_LEFT),
                    anim.value("size", std::vector({1.0, 1.0})),
                    anim.value("requiredLoops", 1)
                );
            }
        }

        return species;
    } catch (const std::exception& e) {
        std::cerr << "Failed to parse species: " << e.what() << std::endl;
        return std::nullopt;
    }
}

File: Sources/windows/App/Sources/Species/SpeciesParser.cs
using Newtonsoft.Json.Linq;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace App
{
    public class SpeciesParser
    {
        private const string SPECIES_ANIMATION_POSITION_FROM_ENTITY_BOTTOM_LEFT = "fromEntityBottomLeft";

        public List<Species> ParseFromFiles(List<String> filePaths) { 
            return filePaths
                .Select(ParseFromFile)
                .Where(x => x != null)
                .ToList();
            }


        public Species? ParseFromFile(string filePath)
        {
            try
            {
                var jsonString = File.ReadAllText(filePath);
                return Parse(jsonString);
            } catch (Exception)
            {
                return null;
            }
        }

        public Species? Parse(String jsonString)
        {
            try
            {
                var json = JObject.Parse(jsonString);
                if (json == null) { return null; }

                var species = new Species(
                    json.GetValue("id")?.ToString() ?? "missingno",
                    (double?)json.GetValue("speed") ?? 1.0,
                    (double?)json.GetValue("scale") ?? 1.0
                );

                species.Capabilities = json.GetValue("capabilities")?.ToObject<List<string>>() ?? new List<string>();
                species.DragPath = json.GetValue("dragPath")?.ToString() ?? "drag";
                species.MovementPath = json.GetValue("movementPath")?.ToString() ?? "walk";
                species.ZIndex = (int?)json.GetValue("zIndex") ?? 0;

                var animations = json.GetValue("animations");
                if (animations != null)
                {
                    foreach (var anim in animations)
                    {
                        if (anim != null) {
                            species.Animations.Add(new SpeciesAnimation(
                                anim["id"]?.ToString() ?? "undefined",
                                anim["position"]?.ToString() ?? SPECIES_ANIMATION_POSITION_FROM_ENTITY_BOTTOM_LEFT,
                                anim["size"]?.ToObject<List<double>>() ?? new List<double> { 1.0, 1.0 },
                                (int?)anim["requiredLoops"] ?? 1
                            ));
                        }
                    }
                }

                return species;
            } catch (Exception e)
            {
                Console.WriteLine($"Failed to parse species: {e.Message}");
                return null;
            }
        }
    }
}
